base_layer/core/src/base_node/sync/block_sync/synchronizer.rs

```diff
@@ -0,0 +1,243 @@
+                "Validating and adding block body for block: #{}", header.height
+            );
+
+            if header.prev_hash != prev_hash {
+                return Err(BlockSyncError::PeerSentBlockThatDidNotFormAChain {
+                    expected: prev_hash.to_hex(),
+                    got: header.prev_hash.to_hex(),
+                });
+            }
+
+            let block_body = block
+                .body
+                .map(AggregateBody::try_from)
+                .ok_or_else(|| BlockSyncError::ReceivedInvalidBlockBody("Block body was empty".to_string()))?
+                .map_err(BlockSyncError::ReceivedInvalidBlockBody)?;
+
+            prev_hash = header.hash();
+
+            debug!(
+                target: LOG_TARGET,
+                "Storing block body #{} (PoW = {}, {} input(s), {} output(s), {} kernel(s))",
+                header.height,
+                header.pow_algo(),
+                block_body.inputs().len(),
+                block_body.outputs().len(),
+                block_body.kernels().len(),
+            );
+
+            let (header, block_body) = self.validate_block(header, block_body).await?;
+
+            self.blockchain_state
+                .write_transaction()
+                .insert_block_body(header_hash.clone(), block_body)
+                .set_metadata(MetadataKey::ChainHeight, MetadataValue::ChainHeight(header.height))
+                .set_metadata(MetadataKey::BestBlock, MetadataValue::BestBlock(header_hash))
+                .commit()
+                .await?;
+
+            debug!(target: LOG_TARGET, "Block body #{} added", header.height);
+            current_header = Some(header);
+        }
+
+        if let Some(header) = current_header {
+            // Update metadata to last tip header
+            let height = header.height;
+            let best_block = header.hash();
+            let accumulated_difficulty = header.get_proof_of_work()?.total_accumulated_difficulty();
+            self.blockchain_state
+                .write_transaction()
+                .set_metadata(MetadataKey::ChainHeight, MetadataValue::ChainHeight(height))
+                .set_metadata(MetadataKey::BestBlock, MetadataValue::BestBlock(best_block.to_vec()))
+                .set_metadata(
+                    MetadataKey::AccumulatedWork,
+                    MetadataValue::AccumulatedWork(accumulated_difficulty),
+                )
+                .commit()
+                .await?;
+        }
+
+        debug!(target: LOG_TARGET, "Completed block sync with peer `{}`", peer);
+
+        Ok(())
+    }
+
+    async fn validate_block(
+        &self,
+        header: BlockHeader,
+        mut body: AggregateBody,
+    ) -> Result<(BlockHeader, AggregateBody), BlockSyncError>
+    {
+        let validator = self.block_validator.clone();
+        task::spawn_blocking(move || {
+            // Ensure that the block body is sorted as required by validation
+            body.sort();
```

Comment:
Rather leave it as is. If the body is not sorted, it should fail validation

Reply:
We don't actually know if it is sorted - that flag is not sent to us, `is_sorted` simply says that `sorted` was called at least once.

Reply:
Though I suppose if it isn't sent in canonical order the MMR check will fail, so this could save us the sort operation. I think that's a good optimisation.